src/optimizer/query_rewrite.c

```diff
@@ -7174,6 +7174,73 @@ qo_move_on_clause_of_explicit_join_to_where_clause (PARSER_CONTEXT * parser, PT_
     }
 }
 
+/*
+ * qo_push_limit_to_union () - push limit on union query
+ *   return: PT_NODE *
+ *   parser(in): parser environment
+ *   node(in): possible query
+ *   arg(in):
+ *   continue_walk(in):
+ */
+static PT_NODE *
+qo_push_limit_to_union (PARSER_CONTEXT * parser, PT_NODE * node, PT_NODE * limit)
+{
+  switch (node->node_type)
+    {
+    case PT_SELECT:
+      if (!pt_has_inst_in_where_and_select_list (parser, node))
+	{
+	  node->info.query.flag.rewrite_limit = 1;
+	  node->info.query.limit = limit;
+	  return node;
+	}
+      break;
+
+    case PT_UNION:
+      qo_push_limit_to_union (parser, node->info.query.q.union_.arg1, limit);
+      qo_push_limit_to_union (parser, node->info.query.q.union_.arg2, limit);
+      break;
+
+    default:
+      break;
+    }
+  return node;
+}
+
+/*
+ * qo_check_distinct_union () - check having distinct in union clause
+ *   return: PT_NODE *
+ *   parser(in): parser environment
+ *   node(in): possible query
+ *   arg(in):
+ *   continue_walk(in):
+ */
+static bool
+qo_check_distinct_union (PARSER_CONTEXT * parser, PT_NODE * node)
+{
+  bool result;
+
+  switch (node->node_type)
+    {
+    case PT_SELECT:
+      break;
+
+    case PT_UNION:
+      if (node->info.query.all_distinct == PT_DISTINCT)
+	{
+	  return true;
+	}
+
+      result |= qo_check_distinct_union (parser, node->info.query.q.union_.arg1);
+      result |= qo_check_distinct_union (parser, node->info.query.q.union_.arg2);
```

Comment:
If any part is true, all will be true.
How about doing something like below?

```
if(qo_check_distinct_union (parser, node->info.query.q.union_.arg1))
  {
    return true;
  }
return qo_check_distinct_union (parser, node->info.query.q.union_.arg2);
```

Reply:
You are right, but I want to keep this code style.